Document all language REPLs and ANSI support

kdeldycke opened the pull request:
This PR:
- lists all missing interactive sessions from langage REPL
- groups sessions in OS shells or langage REPLs
- documents the generic output lexer 
- adds session examples
- documents the lack of ANSI support in output
- adds pointers to third-party projects for ANSI support

Relates to:
- #1148
- #1422
- #477

kdeldycke commented on doc/docs/terminal-sessions.rst:
```diff
@@ -21,26 +32,288 @@ prompt.
-     PS> Get-ChildItem
-     >> Get-ChildItem
+Similarly to systems shells, Pygments recognize a variety of interactive
+language sessions:
```
OK, I just moved all the examples to the docstrings.

kdeldycke commented:
This PR is ready to be reviewed and merged, even if the CI job on 3.12-dev is failing. See: https://github.com/pygments/pygments/pull/2445

Anteru commented:
Thanks! I'm a bit worried that the examples in the doc string will actually work with the lexer .. I'd prefer them to be also committed as test snippets. At that point however, we probably shouldn't duplicate them from the test snippets into the documentation. This could be something to solve via a private class field for the documentation, maybe something like `_doc_example = '/path/to/example'` which is then parsed by the lexer page generator. This would have to go here: https://github.com/pygments/pygments/blob/master/pygments/sphinxext.py#L148

This may sound like overkill, but I think that's the best long term solution to make sure those things don't go out of sync, and it's probably (long term) a good idea to have some example file linked to every lexer.

kdeldycke commented:
Thanks @Anteru for the detailed feedback!

I have these kind of reflexion too with my [Click Extra](https://github.com/kdeldycke/click-extra) project, in which I have to maintain some code to keep the document and code in sync.

I have several strategy for that:
- [dedicated Sphinx extension](https://kdeldycke.github.io/click-extra/sphinx.html) to produce Click code and invokation example
- stringent [unittests to check the metadata and content of third-party dependencies](https://github.com/kdeldycke/click-extra/blob/main/click_extra/tests/test_pygments.py#L58-L145) (including Pygments itself) 
- regular runs via GitHub workflows to [automatically produce dynamic documentation](https://github.com/kdeldycke/click-extra/blob/b5c456254378056160c6c2d541e38dba35e53246/click_extra/docs_update.py#L70-L97) and push a PR

It's a little bit all over the place by does the job of reducing the cognitive work of maintenance. I'm just pointing to my stuff as inspiration: it might not be adapted to the nature of Pygments. But I guess I know enough about Pygments and Sphinx to attempt an implementation of the long term solution you described.

But should this be done in this PR? I don't think so because of the bigger scope of that feature: it would introduce new code while this PR is only introducing rST content. So what's missing for this PR to be considered for a merge upstream?

Anteru commented:
Honestly, I'm not sure if the added cost of having to convert this later is worth to merge this now (I'm really afraid of code rot quickly.) I'll see if I can spare some cycles to convert this, maybe it's not as complicated as I'm afraid it is. I'm not a huge fan of having this extra field on the lexer class to point to a document in the test tree, so the other alternative is to have tests somehow marked up as examples that should show up in the doc for that particular lexer. Let me think a bit more about this, and at the very least try to prototype the `_example` thing ... I did some work on the Sphinx extension and it feels like this shouldn't be that hard other than having to somehow pipe it through Pygments while building the docs and _then_ insert it back into the doctree.

kdeldycke commented:
Since your last comment @Anteru, I realized you went ahead with the `_example` attribute and implemented it upstream in #2451.

So I updated this PR to use it to document all REPL sessions. All tests are passing, this PR is ready to be merged upstream! :)

Anteru commented on doc/docs/terminal-sessions.rst:
```diff
@@ -1,46 +1,55 @@
-brackets ``[any]`` represent optional parts of the prompt, and prompt parts
-without brackets or in parenthesis ``(any)`` represent required parts of the
-prompt.
+They are typiccaly command lines or code, mixed with generic output.
```
```suggestion
They are typically command lines or code, mixed with generic output.
```

Anteru commented on doc/docs/terminal-sessions.rst:
```diff
@@ -1,46 +1,55 @@
+shell session, prefix your code snippet with a specially formatted prompt.
 
-  .. code-block:: doscon
+They are typiccaly named ``<shell> Session``.
```
```suggestion
They are typically named ``<shell> Session``.
```

Anteru commented on doc/docs/terminal-sessions.rst:
```diff
@@ -1,46 +1,55 @@
 
-     (any)> ls -lh
-     ? ls -lh
+Their IDs are typiccaly follow the ``<language>-console`` or
```
```suggestion
Their IDs typically follow the ``<language>-console`` or
```

Anteru commented on doc/docs/terminal-sessions.rst:
```diff
@@ -1,46 +1,55 @@
+--------------
+
+In all the lexers above, the command results are parsed as generic output.
+Which means they are rendered as-is, without any tentative of rendering
+the styling carried on by ANSI codes.
```
```suggestion
Which means they are rendered as-is, without any styling applied, for example by ANSI codes.
```

kdeldycke commented:
@Anteru I just updated the PR with your requested changes, and squashed the commits. All tests are passing, this PR is ready to be merged upstream.

Anteru commented:
Thanks!
